### PR TITLE
Update package infrastructure

### PR DESCRIPTION
**Pull request by Cadair**
This Pull Request does a few things:

1) Configure Github releases as changelog. This configures the GitHub automatic release notes generator and a workflow to copy the release notes into the changelog when a release is made.
2) Add a config for pre-commit (lightly modified from the one in astropy core).
3) Update the config for isort
4) run pyupgrade
5) Remove `setup.py` and configure all the packaging in `setup.cfg` and `pyproject.toml`.

### CODE REVIEW

**Comment by Cadair** [comment]
@astrofrog I think I have finally made this pass.

File: RELEASE_INSTRUCTIONS.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+As a mantainer it should be nice and simple to do, especially if all merged PRs
+have nice titles and are correctly labelled.
+
+Here is the process to follow to make a new release:
+
+* Go through all the PRs since the last release, make sure they have
+  descriptive titles (these will become the changelog entry) and are labelled
+  correctly.
+* Go to the GitHub releases interface and draft a new release, tags should drop
+  the trailing `.0` for historical reasons on major releases.
```
Comment (astrofrog) [inline RELEASE_INSTRUCTIONS.md:14]: Why don't we just switch to semantic versioning and no longer remove the 0?

**Review by astrofrog — APPROVED** [review]
Thank you for doing all this! Just a small comment below but feel free to merge otherwise.

File: RELEASE_INSTRUCTIONS.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+As a mantainer it should be nice and simple to do, especially if all merged PRs
+have nice titles and are correctly labelled.
+
+Here is the process to follow to make a new release:
+
+* Go through all the PRs since the last release, make sure they have
+  descriptive titles (these will become the changelog entry) and are labelled
+  correctly.
+* Go to the GitHub releases interface and draft a new release, new tags should
+  drop include the trailing `.0` on major releases. (Releases prior to 0.10.0
```
Comment (astrofrog) [inline RELEASE_INSTRUCTIONS.md:14]: Remove drop

Comment (Cadair) [reply RELEASE_INSTRUCTIONS.md:14]:
```suggestion
  include the trailing `.0` on major releases. (Releases prior to 0.10.0
```

**Comment by bsipocz** [comment]
FYI, this PR took away `reproject.test()` without mentioning it in the changelog, etc.

**Comment by astrofrog** [comment]
@bsipocz Thanks for spotting this, don't think it was deliberate. How did you notice btw?

**Comment by bsipocz** [comment]
conda-forge packaging did run the user tests

**Comment by astrofrog** [comment]
Ah I see! I will check how easy it is to add back and do a patch release

**Pull request merged**